### PR TITLE
Fix bugs in RefineMesh Event

### PR DESCRIPTION
**Pull request by knelli2**
## Proposed changes

Also adds an option in the input file to choose whether the code should error if it tries to refine beyond the limits (but is stopped by enforcing the limits) since this could be an indicator of something going wrong. Currently, I set it to *not* error in all the executables to keep the original behavior. If we decide this is something we want to turn on in certain executables, we can do so later.

### Upgrade instructions

Must add a new option `ErrorBeyondLimits: False` to the AMR limits section, i.e.

```yaml
Amr:
  ...
  Policies:
    ...
    Limits:
      ...
      ErrorBeyondLimits: False
```

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### CODE REVIEW

File: src/ParallelAlgorithms/Amr/Policies/Limits.hpp [hunk 1]
```diff
@@ -58,7 +58,20 @@ class Limits {
...
+  /// \brief Whether the code should error if EnforcePolicies tries has to
+  /// prevent refinement from going beyond the given limits.
+  ///
+  /// \details The Limits class is just a holder for this value, the actual
+  /// error happens in `amr::Actions::EvaluateRefinementCriteria` or
+  /// `amr::Events::RefineMesh`
+  struct ErrorBeyondLimits {
+    using type = bool;
+    static constexpr Options::String help = {
+        "If p-refinement tries to go beyind the RefinementLevel or "
```
Comment (kidder) [inline src/ParallelAlgorithms/Amr/Policies/Limits.hpp]: p-refinement -> adaptive mesh refinement
beyind -> beyond

**Review by kidder — CHANGES_REQUESTED** [review]
you can squash the fix

File: src/ParallelAlgorithms/Amr/Policies/Limits.hpp (hunk 1 above)
Comment (knelli2) [reply src/ParallelAlgorithms/Amr/Policies/Limits.hpp]: Done!

File: src/ParallelAlgorithms/Amr/Policies/EnforcePolicies.hpp [hunk 2]
```diff
@@ -31,8 +34,14 @@ namespace amr {
 ///   their bounds, the Flag is changed to DoNothing.  Note that the minimum
 ///   resolution is Basis and Quadrature dependent, so may be higher than the
 ///   given amr_policies.refinement_limits().minimum_resolution()
+/// - If the `amr_decision` tried to go beyond the `amr::Limits` (and was
+///   corrected by the `amr_policies`), then this returns the
+///   `amr::Limits::LimitType`s that were obeyed in each direction. Otherwsie,
```
Comment (nilsvu) [inline src/ParallelAlgorithms/Amr/Policies/EnforcePolicies.hpp]: typo "otherwise"

File: src/ParallelAlgorithms/Amr/Policies/Limits.hpp [hunk 3]
```diff
@@ -28,7 +28,7 @@ namespace amr {
 ///   resolution is bounded between 1 and
 ///   Spectral::maximum_number_of_points<Spectral::Basis::Legendre>
 ///   which are limits based on the implementation details of ElementId and
-///   Mesh.
+///   Mesh. There is no error for attemtping to go beyind the limits.
```
Comment (nilsvu) [inline src/ParallelAlgorithms/Amr/Policies/Limits.hpp]: typo "attempting", "beyond"

File: src/ParallelAlgorithms/Amr/Actions/EvaluateRefinementCriteria.hpp [hunk 4]
```diff
@@ -108,9 +108,21 @@ struct EvaluateRefinementCriteria {
...
+    const auto limit_types = amr::enforce_policies(
+        make_not_null(&overall_decision), policies, element_id,
+        get<::domain::Tags::Mesh<volume_dim>>(box));
+    if (policies.limits().error_beyond_limits() and
+        alg::any_of(
+            limit_types,
+            [](const std::optional<amr::Limits::LimitType>& limit_type) {
+              return limit_type.has_value();
+            })) {
+      ERROR("Tried refining beyond the AMR limits in element "
```
Comment (nilsvu) [inline src/ParallelAlgorithms/Amr/Actions/EvaluateRefinementCriteria.hpp]: Throw this error inside `amr::enforce_policies`. Then you don't need to return anything, don't need the enum, don't have to repeat the error, and it just seems easier overall.

File: docs/DevGuide/Amr.md [hunk 5]
```diff
@@ -343,11 +343,14 @@ Amr:
...
   Verbosity: Verbose
 ```
 
 Note that the values `Auto` for the Limits options choose the default
-limits set by the code.
+limits set by the code. Also note that if a criteria tries to refine beyond the
+limits, whether or not the code should error is controlled by
+`ErrorBeyondLimits`.
 
 "When" AMR happens is contolled by specifying a Trigger and a list of
```
Comment (nilsvu) [inline docs/DevGuide/Amr.md]: typo "controlled" (not yours but might as well fix)

**Review by nilsvu — CHANGES_REQUESTED** [review]

**Comment by knelli2** [comment]
@nilsvu That simplified things a bit

File: docs/DevGuide/Amr.md [hunk 6]
```diff
@@ -343,13 +343,16 @@ Amr:
...
 ```
 
 Note that the values `Auto` for the Limits options choose the default
-limits set by the code.
+limits set by the code. Also note that if a criteria tries to refine beyond the
+limits, whether or not the code should error is controlled by
+`ErrorBeyondLimits`.
 
-"When" AMR happens is contolled by specifying a Trigger and a list of
+"When" AMR happens is conrtolled by specifying a Trigger and a list of
```
Comment (nilsvu) [inline docs/DevGuide/Amr.md]: oops :D

File: docs/DevGuide/Amr.md [hunk 7]
```diff
@@ -343,13 +343,16 @@ Amr:
     Limits:
       RefinementLevel: Auto
       NumGridPoints: Auto
+      ErrorBeyondLimits: False
   Verbosity: Verbose
 ```
 
 Note that the values `Auto` for the Limits options choose the default
-limits set by the code.
+limits set by the code. Also note that if a criteria tries to refine beyond the
```
Comment (nilsvu) [inline docs/DevGuide/Amr.md]: "criteria try"

File: src/ParallelAlgorithms/Amr/Policies/EnforcePolicies.hpp [hunk 8]
```diff
@@ -31,6 +31,11 @@ namespace amr {
 ///   their bounds, the Flag is changed to DoNothing.  Note that the minimum
 ///   resolution is Basis and Quadrature dependent, so may be higher than the
 ///   given amr_policies.refinement_limits().minimum_resolution()
+/// - If the `amr_decision` tried to go beyond the `amr::Limits` (and was
+///   corrected by the `amr_policies`), then this returns the
+///   `amr::Limits::LimitType`s that were obeyed in each direction. Otherwise,
```
Comment (nilsvu) [inline src/ParallelAlgorithms/Amr/Policies/EnforcePolicies.hpp]: update docs

**Review by nilsvu — CHANGES_REQUESTED** [review]

File: docs/DevGuide/Amr.md (hunk 6 above)
Comment (knelli2) [reply docs/DevGuide/Amr.md]: Oops...

File: src/ParallelAlgorithms/Amr/Policies/Limits.hpp [hunk 9]
```diff
@@ -6,6 +6,7 @@
 #include <array>
 #include <cstddef>
 #include <optional>
+#include <ostream>
```
Comment (kidder) [inline src/ParallelAlgorithms/Amr/Policies/Limits.hpp]: don't need this include

File: src/ParallelAlgorithms/Amr/Policies/Limits.cpp [hunk 10]
```diff
@@ -3,6 +3,7 @@
 
 #include "ParallelAlgorithms/Amr/Policies/Limits.hpp"
 
+#include <ostream>
```
Comment (kidder) [inline src/ParallelAlgorithms/Amr/Policies/Limits.cpp]: don't need this include

**Review by kidder — CHANGES_REQUESTED** [review]
you can squash

**Review by kidder — APPROVED** [review]

**Review by nilsvu — APPROVED** [review]

**Pull request merged**